### DISCUSSION TITLE
Should we use mpremote or pyboard to test

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#10282] [Using MicroPython] [answered]
Hello fellow MicroPython enthousiasts,
I'm trying to figure out what would be the best approach to do unittests or integration tests of MicroPython code while a hardware device is connected. Any insights are most welcome!

Using unittest and unittest-discover for MicroPython seems to work ok for simple code testing using stubs or mocking things related to the hardware. But is there a good way to for example use pytest (or the untittest) module of CPython and write tests that can also executes code on the connected board to validate a hardware test outcome?

* I know there is `pyboard.py` that is capable of executing code on a connected board as it is an importable module in CPython?
* Or should we use `mpremote`? This seems to be a command-line tool only so we would need to execute hardware stuff using mpremote running inside a CPython subprocess for this...

Do we know what the roadmap is for `mpremote`? Is this the one that will replace `pyboard.py` or will it be an importable module in the future... or do we just use `pyboard.py` for this?

Thanks!

**Comment by jimmo** [comment, marked as answer]
`pyboard.py` is a library for interacting with connected boards, as well as a rudimentary command line interface.
`mpremote` uses pyboard.py in order to provide a much more complete command line interface.

If you install mpremote from pip, you also get pyboard, so even if you just want to remote control your board from Python, you can make mpremote a project dependency and just use the pyboard module.

I do not recommend using mpremote via subprocess! Everything you need should be available from pyboard.py.

FWIW, I just sent a PR to add some additional functionality that might be useful when using pyboard.py as a library. https://github.com/micropython/micropython/pull/10286

**Reply by saraverbeecke** [reply]
@jimmo great work again! Thanks!

**Comment by ubidefeo** [comment]
@jimmo 
could you please clarify the following?
>If you install mpremote from pip, you also get pyboard

I have installed `mpremote` but if I enter a Python session and try to import `pyboard` I get `ModuleNotFoundError`.
Unfortunately `pip install pyboard` will install something which seems unsupported since 2019.

At the moment I manually brought `pyboard.py` from the MicroPython repo into a project, but it feels beyond wrong.
What's the solution to installing `pyboard.py`?

**Reply by Josverl** [reply]
`from mpremote import pyboard` should work

**Reply by jimmo** [reply]
Yes, what @Josverl said. Because mpremote internally uses pyboard.py it ships this module as part of the PyPI wheel.

However, in the next release of mpremote this will no longer be the case as we are developing a more comprehensive Python API for controlling devices. This will still be part of mpremote, but it will no longer be the `pyboard` module. In other words, the `mpremote` PyPI package will be both the command line tool and the Python scripting interface.

This is a bit of a WIP at the moment. I agree it's not great, but for now if you just want to use the `pyboard` module, copying it from the repo is probably the best option for now, and then in the future you can migrate your code to the new mpremote API.

**Reply by ubidefeo** [reply]
thanks @Josverl and @jimmo 

>the mpremote PyPI package will be both the command line tool and the Python scripting interface.

I like this a lot, and I love how lively MicroPython is.
At Arduino I'm pushing a lot to spread it, because I absolutely love it for my classes :)

u.
